File: python/data_planning_service.py
import mparticle
import json
import pdb
import requests
import sys

from mparticle import *
from time import sleep, time


class DataPlanningService:
    platform_api_url = 'https://api.mparticle.com'
    mparticle_sso_token_endpoint = 'https://sso.auth.mparticle.com/oauth/token'

    def __init__(self, config_json):
        with open(config_json) as f:
            config = json.load(f)

        self.platform_api_key = config['platform_api_key']
        self.platform_api_secret = config['platform_api_secret']

        data_json = {
            "client_id": self.platform_api_key,
            "client_secret": self.platform_api_secret,
            "audience": self.platform_api_url,
            "grant_type": "client_credentials"
        }
        
        oauth_response = requests.post(self.mparticle_sso_token_endpoint, data=data_json)
        if oauth_response.status_code != 200:
            print(f'Authentication request returned {oauth_response.status_code}')
            sys.exit(1)
    
        oauth_response_json = json.loads(oauth_response.text)
        self.headers = {'Authorization': 'Bearer ' + oauth_response_json['access_token']}

    def create_data_plan(self, workspace_id, fname):
        with open(fname) as f:
            data_plan = json.load(f)
        path = '/platform/v2/workspaces/%d/plans' % (workspace_id)
        url = self.platform_api_url + path
        response = requests.post(url, json=data_plan, headers=self.headers)
        if (response.status_code > 200):
            print(response.json())
        response.raise_for_status()
        plan = response.json()
        return plan

    def download_data_plans(self, workspace_id):
        path = '/platform/v2/workspaces/%d/plans' % (workspace_id)
        url = self.platform_api_url + path
        response = requests.get(url, headers=self.headers)
        if (response.status_code > 200):
            print(response.json())
        response.raise_for_status()
        plans = response.json()
        return plans
    
    def download_data_plan(self, workspace_id, plan_id):
        path = '/platform/v2/workspaces/%d/plans/%s' % (workspace_id, plan_id)
        url = self.platform_api_url + path
        response = requests.get(url, headers=self.headers)
        if (response.status_code > 200):
            print(response.json())
        response.raise_for_status()
        plan_version = response.json()
        return plan_version

    def update_data_plan(self, workspace_id, plan_id, fname):
        with open(fname) as f:
            data_plan = json.load(f)
            
        path = '/platform/v2/workspaces/%d/plans/%s' % (workspace_id, plan_id)
        url = self.platform_api_url + path
        response = requests.patch(url, headers=self.headers, json=data_plan)
        if (response.status_code > 200):
            print(response.json())
        response.raise_for_status()
        plan = response.json()
        return plan
    
    def delete_data_plan(self, workspace_id, plan_id):            
        path = '/platform/v2/workspaces/%d/plans/%s' % (workspace_id, plan_id)
        url = self.platform_api_url + path
        response = requests.delete(url, headers=self.headers)
        if (response.status_code > 204):
            print(response.json())
        response.raise_for_status()
    
    def download_data_plan_version(self, workspace_id, plan_id, plan_version):
        path = '/platform/v2/workspaces/%d/plans/%s/versions/%s' % (workspace_id, plan_id, plan_version)
        url = self.platform_api_url + path
        response = requests.get(url, headers=self.headers)
        if (response.status_code > 200):
            print(response.json())
        response.raise_for_status()
        plan_version = response.json()
        return plan_version

    def create_data_plan_version(self, workspace_id, plan_id, fname):
        with open(fname) as f:
            version = json.load(f)

        path = '/platform/v2/workspaces/%d/plans/%s/versions' % (workspace_id, plan_id)
        url = self.platform_api_url + path
        response = requests.post(url, headers=self.headers, json=version)
        if (response.status_code > 200):
            print(response.json())
        response.raise_for_status()
        plan_version = response.json()
        return plan_version
    
    def update_data_plan_version(self, workspace_id, plan_id, plan_version, fname):
        with open(fname) as f:
            version = json.load(f)

        path = '/platform/v2/workspaces/%d/plans/%s/versions/%s' % (workspace_id, plan_id, plan_version)
        url = self.platform_api_url + path
        response = requests.patch(url, headers=self.headers, json=version)
        responseJson = response.json()
        if (response.status_code > 200):
            print(responseJson)
        response.raise_for_status()
        return responseJson

    def delete_data_plan_version(self, workspace_id, plan_id, plan_version):
        path = '/platform/v2/workspaces/%d/plans/%s/versions/%s' % (workspace_id, plan_id, plan_version)
        url = self.platform_api_url + path
        response = requests.delete(url, headers=self.headers)
        if (response.status_code > 204):
            print(response.json())
        response.raise_for_status()
    
    def get_latest_plan_version(self, workspace_id, plan_id):
        path = '/platform/v2/workspaces/%d/plans/%s' % (workspace_id, plan_id)
        url = self.platform_api_url + path
        response = requests.get(url, headers=self.headers)
        if (response.status_code > 200):
            print(response.json())
        response.raise_for_status()
        plan_versions = response.json()
        return max([plan_version['version'] for plan_version in plan_versions['data_plan_versions']])

File: nodejs/data_planning_service.js
'use strict';

const axios = require('axios');
const fs = require('fs');

module.exports = class DataPlanningService {
    platform_api_url = 'https://api.mparticle.com';
    mparticle_sso_token_endpoint = 'https://sso.auth.mparticle.com/oauth/token';

    init(config_json) {
        const config = JSON.parse(fs.readFileSync(config_json));

        this.platform_api_key = config['platform_api_key'];
        this.platform_api_secret = config['platform_api_secret'];

        const data_json = {
            "client_id": this.platform_api_key,
            "client_secret": this.platform_api_secret,
            "audience": this.platform_api_url,
            "grant_type": "client_credentials",
        };
        
        return axios
            .post(this.mparticle_sso_token_endpoint, data_json)
            .then(res => {
                axios.defaults.headers.common['Authorization'] = `Bearer ${res.data['access_token']}`;
            });
    }

    create_data_plan(workspace_id, fname) {
        const data_plan = JSON.parse(fs.readFileSync(fname));
        const url = `${this.platform_api_url}/platform/v2/workspaces/${workspace_id}/plans`;

        return axios.post(url, data_plan);
    }

    download_data_plans(workspace_id) {
        const url = `${this.platform_api_url}/platform/v2/workspaces/${workspace_id}/plans`;
        return axios.get(url);
    }
    
    download_data_plan(workspace_id, plan_id) {
        const url = `${this.platform_api_url}/platform/v2/workspaces/${workspace_id}/plans/${plan_id}`;
        return axios.get(url);
    }

    update_data_plan(workspace_id, plan_id, fname) {
        const data_plan = JSON.parse(fs.readFileSync(fname));
            
        const url = `${this.platform_api_url}/platform/v2/workspaces/${workspace_id}/plans/${plan_id}`;
        return axios.patch(url, data_plan);
    }
    
    delete_data_plan(workspace_id, plan_id) {         
        const url = `${this.platform_api_url}/platform/v2/workspaces/${workspace_id}/plans/${plan_id}`;
        return axios.delete(url);
    }
    
    download_data_plan_version(workspace_id, plan_id, plan_version) {
        const url = `${this.platform_api_url}/platform/v2/workspaces/${workspace_id}/plans/${plan_id}/versions/${plan_version}`;
        return axios.get(url);
    }

    create_data_plan_version(workspace_id, plan_id, fname) {
        const version = JSON.parse(fs.readFileSync(fname));

        const url = `${this.platform_api_url}/platform/v2/workspaces/${workspace_id}/plans/${plan_id}/versions`;
        return axios.post(url, version);
    }
    
    update_data_plan_version(workspace_id, plan_id, plan_version, fname) {
        const version = JSON.parse(fs.readFileSync(fname));

        const url = `${this.platform_api_url}/platform/v2/workspaces/${workspace_id}/plans/${plan_id}/versions/${plan_version}`;
        return axios.patch(url, version);
    }

    delete_data_plan_version(workspace_id, plan_id, plan_version) {
        const url = `${this.platform_api_url}/platform/v2/workspaces/${workspace_id}/plans/${plan_id}/versions/${plan_version}`;
        return axios.delete(url);
    }

    validate_batch(workspace_id, data_plan_file, batch_file) {
        const document = JSON.parse(fs.readFileSync(data_plan_file));
        const batch = JSON.parse(fs.readFileSync(batch_file));

        const payload = {
            "document": document["version_document"],
            "batch": batch,
        };

        const url = `${this.platform_api_url}/platform/v2/workspaces/${workspace_id}/plans/validate`;
        return axios.post(url, payload);
    }
}
